File: magmac/src/com/meti/api/collect/List.java
package com.meti.api.collect;

import com.meti.api.iterator.Iterator;
import com.meti.api.option.Option;

public interface List<T> {
    Option<T> head();

    boolean isEmpty();

    Iterator<T> iter();

    List<T> add(T element);

    int size();

    Option<T> get(int index);
}

File: src/java/magma/api/collect/List.ts
import from 'magma.api.option.Option'.ts;
import from 'magma.app.Main'.ts;
interface List<T> {
	add(element: T): List<T>;
	query(): Main.Query<T>;
	size(): number;
	subList(startInclusive: number, endExclusive: number): Option<List<T>>;
	findLast(): Option<T>;
	findFirst(): Option<T>;
	find(index: number): Option<T>;
	queryWithIndices(): Main.Query<Main.Tuple<number, T>>;
	addAll(others: List<T>): List<T>;
	contains(element: T): boolean;
	queryReversed(): Main.Query<T>;
}
